### ISSUE TITLE
Custom hover for CSS color with ColorPicker

### ISSUE THREAD

**Issue by angelozerr** [mickaelistria/eclipse-bluesky#41]
Once I will finish https://github.com/mickaelistria/eclipse-bluesky/issues/30 I would like to provide a ColorPicker (like VSCode does) to choose color.

@mickaelistria do you think it's possible to provide custom hover with GenericEditor? If yes, could you give me some directive please to implement custom hover. Thanks!

**Comment by mickaelistria** [comment]
Yes, just implement the ITextHover, ITextHoverExtension, ITextHoverExtension2 interfaces, and in this implementation, make the IInformationControlCreator create the right widgets. Then associate your hover class with the CSS content-type and it should enable it.

**Comment by angelozerr** [comment]
Thanks @mickaelistria, I will try it.

**Comment by angelozerr** [comment]
Here a little demo of my current work about CSS Color hover:

![colorpickerdemo](https://user-images.githubusercontent.com/1932211/30371961-854d3662-987c-11e7-940e-901bc781d395.gif)

I'm waiting a fix for problem which closes hover when color picker is opened https://bugs.eclipse.org/bugs/show_bug.cgi?id=522203#c4 and after I will create a PR

**Comment by mickaelistria** [comment]
Why is this an issue to close the hover popup? I would expect click "OK" on
the color picker to directly apply the change. I don't have the impression
need to have the hover popup persisting behind the color picker.
In general, I think it's in the very root of Hover mechanism in JFace to
hide it as soon as focus changes ; so I'm afraid it's going to be hard to
change that.
Instead of a button there, you could consider using a link, as the action
of clicking the button itself doesn't really have an immediate effect on
the file.

**Comment by angelozerr** [comment]
> I would expect click "OK" on the color picker to directly apply the change.

You mean having the Color picker UI in the hover? If it that I totally agree with you, but not possible because ColorDialog uses native color picker and works only in a dialog. Developping my own color picker is a thing that I would like to avoid (I'm afraid to spend a lot of time to do that).

> In general, I think it's in the very root of Hover mechanism in JFace to
hide it as soon as focus changes ; so I'm afraid it's going to be hard to
change that.

-(

> Instead of a button there, you could consider using a link, as the action of clicking the button itself doesn't really have an immediate effect on the file.

Sorry I don't understand?

**Comment by mickaelistria** [comment]
You wrote "I'm waiting a fix for problem which closes hover when color
picker is opened". My question is "why is it a problem?" ie if hover closes
and color picker shows up, it seems to be an acceptable behavior.

**Comment by angelozerr** [comment]
> My question is "why is it a problem?" ie if hover closes and color picker shows up, it seems to be an acceptable behavior.

The big problem is that if hover close while ColorDialog is displayed, it crashes the Eclipse IDE.

**Comment by mickaelistria** [comment]
>
> The big problem is that if hover close while ColorDialog is displayed, it
> crashes the Eclipse IDE.
>
Have you tried using another parent shell? Instead of using the Hover
shell, this dialog would probably better use the editor's active shell as
parent (because once it's opened, it's more related to the editor than to
the hover actually).

**Comment by angelozerr** [comment]
> Have you tried using another parent shell? 

Good catch! I'm waiting for accepting PR https://github.com/mickaelistria/eclipse-bluesky/pull/43 to do a new PR with color picker.

**Comment by mickaelistria** [comment]
What's in #43 that's necessary to implement the color picker?

**Comment by angelozerr** [comment]
I'm using ColorSymbolAnnotation to retrieve if there are color or not and activate hover.

**Comment by angelozerr** [comment]
@mickaelistria @mniewrzal I have an another idea to choose color: I do that without hover but just by clicking on color square. Here a little demo:

![simplecolorpickerdemo](https://user-images.githubusercontent.com/1932211/30434012-5a3210d4-9966-11e7-87e5-7fbb87971c25.gif)

 * Pros: with just one click you select your color.
 * Cons: you cannot choose format rgb, hexa color. Where can I add that, in an hover?

Please give me your feedback. Thanks!

**Comment by mickaelistria** [comment]
I think having it on both click and hover makes sense.
Click is nicer for people using touchscreens (like Surface Pro machines) as
touchscreens don't encourage much hovers.
Hover is usually easier to discover for users having a mouse because the
features enable itself automatically.

**Comment by angelozerr** [comment]
Ok so if I understand you wish to have both hover and click.

I propose you to create 2 PRs:

 * one with click
 * one with hover

Having 2 PRs will ease review. Are you OK?

**Comment by mickaelistria** [comment]
ok!

**Comment by angelozerr** [comment]
@mickaelistria @ymniewrzal you can find PR with color dialog opened by click at https://github.com/mickaelistria/eclipse-bluesky/pull/48
